### PR TITLE
[with findfont diff] Parsing all families in font_manager

### PR DESCRIPTION
**Pull request by aitikgupta**
#### This is the beginning of migrating from Matplotlib's "Font-First" approach to a "Text-First" approach.

The very first step is to parse all families in font manager. Previously, we only parsed families until we find a font file, and the rest of the backends are accustomed to just a single font file (which needs changing)

I'm _hoping_ this breaks many tests, so as to know exactly what to fix at other places.

- Change the return signature of `findfont`, from simple 'str' to 'OrderedDict'
- This OrderedDict would contain fonts defined by `font.family`
    - which are present on the system
    - but also append "fallback fonts" in the end. (which is triggered if a certain font family is not found in the system)
 
I've added a bunch of To-Dos (future PRs) and some fixes at some places.
## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by aitikgupta** [comment]
The tests now fail only due to an unrelated issue, and seems to be affecting all recent PRs.
```console
FAILED lib/matplotlib/tests/test_image.py::test_huge_range_log[png] - ValueError: Invalid vmin or vmax
```

File: examples/misc/logos2.py [hunk 1]
```diff
@@ -20,10 +20,10 @@
 def get_font_properties():
     # The original font is Calibri, if that is not installed, we fall back
     # to Carlito, which is metrically equivalent.
-    if 'Calibri' in matplotlib.font_manager.findfont('Calibri:bold'):
+    if 'Calibri' in matplotlib.font_manager.findfont('Calibri:bold').keys():
```
Comment (jklymak) [inline examples/misc/logos2.py:23]: I don't think you need `.keys()`?

**Review by jklymak — COMMENTED** [review]
`findfont` is public API, so can we just change its return type?  At the very least this requires an API change note?

File: examples/misc/logos2.py (hunk 1 above)
Comment (aitikgupta) [reply examples/misc/logos2.py:23]: ah that's right, I was probably searching for findfont usages within mpl and happen to change this eagerly.

**Comment by aitikgupta** [comment]
> `findfont` is public API

_Because_ this is a public API, the return type change is quite huge, which would break things at a lot of places.
I've proposed a 'without findfont diff' parsing of all families, which would only change internals of mpl and won't affect end users. see: https://github.com/matplotlib/matplotlib/pull/20549

I think this PR should still be open for a while until we've got a good look at the other PR's possibilities.

**Comment by jklymak** [comment]
Please reference other relevant PRs in a PR before asking folks to review.  I had no idea there were two, and we probably should decide on an approach before reviewing.

**Comment by aitikgupta** [comment]
> Please reference other relevant PRs in a PR before asking folks to review.

I did? I see a ```aitikgupta mentioned this pull request 5 days ago``` just above your review @jklymak
(if that's the referencing you meant?) there's also detail about this in the new PR in the ```Short Note```.
Though I apologise if I missed anything.. 😅 

On the other hand, the logic still needs to be reviewed (which is mostly similar in this and the other PR), which is also why I didn't convert it into a draft before..

**Comment by jklymak** [comment]
Folks aren't going to click through every link on a PR.  A quick comment here that this was likely superseded by the other PR would have been helpful.

**Comment by tacaswell** [comment]
I am going to close this PR because the core of it got merged in #20740 and we have decided to keep the API on `findfont` unchanged.

**Pull request closed without merging**